### PR TITLE
Add value conversion methods for complex types

### PR DESCRIPTION
**Pull request by mpenick**
Add methods to convert from `Value` to Java types.

Also, add value conversion methods for less complex types for
completeness.

This doesn't include collections, UDTs or tuples. That's a PR in itself.

**Checklist**
- [X ] Changes manually tested
- [X] Automated Tests added/updated
- [ ] Documentation added/updated

### CODE REVIEW

File: grpc-proto/src/test/java/ValuesTest.java [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+import java.nio.ByteBuffer;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.UUID;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+public class ValuesTest {
+
+  @Nested
```
Comment (tatu-at-datastax) [inline grpc-proto/src/test/java/ValuesTest.java:18]: Quick question on this: I have seen use of `@Nested`, but was curious as to its benefits -- is it just for improved grouping of test results, or such? Nothing against the use just want to understand benefits over either including tests directly in main test class, or as separate main-level test cases.

File: grpc-proto/src/test/java/ValuesTest.java [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  class IntTest {
+
+    @Test
+    public void encodeDecode() {
+      int expected = Integer.MAX_VALUE;
+      assertThat(Values.int_(Values.of(expected))).isEqualTo(expected);
+    }
+
+    @Test
+    public void invalid() {
```
Comment (tatu-at-datastax) [inline grpc-proto/src/test/java/ValuesTest.java:47]: Would it also be possible to test value overflow, since it is checked by converter?

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 3]
```diff
@@ -163,4 +165,143 @@ public static Value of(Map<Value, Value> fields) {
...
+    }
+    return value.getBoolean();
+  }
+
+  public static int int_(Value value) {
+    if (value.getInnerCase() != InnerCase.INT) {
+      throw new IllegalArgumentException("Expected int value");
+    }
+    int intValue = (int) value.getInt();
+    if (intValue != value.getInt()) {
```
Comment (tatu-at-datastax) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:181]: Would this actually fail? As opposed to using same coercion here as in above line.
(I did add a note to test this case to make sure)

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 4]
```diff
@@ -163,4 +165,143 @@ public static Value of(Map<Value, Value> fields) {
...
+
+  public static float float_(Value value) {
+    if (value.getInnerCase() != InnerCase.FLOAT) {
+      throw new IllegalArgumentException("Expected float value");
+    }
+    return value.getFloat();
+  }
+
+  public static double double_(Value value) {
+    if (value.getInnerCase() != InnerCase.DOUBLE) {
```
Comment (tatu-at-datastax) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:227]: This is probably correct but just want to verify: should it not coerce from `FLOAT` to `DOUBLE`?

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 5]
```diff
@@ -163,4 +165,143 @@ public static Value of(Map<Value, Value> fields) {
...
+      throw new IllegalArgumentException("Expected double value");
+    }
+    return value.getDouble();
+  }
+
+  public static ByteBuffer byteBuffer(Value value) {
+    if (value.getInnerCase() != InnerCase.BYTES) {
+      throw new IllegalArgumentException("Expected bytes value");
+    }
+    ByteBuffer bytes = ByteBuffer.allocate(value.getBytes().size());
```
Comment (tatu-at-datastax) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:237]: Does `getBytes()` always return a new copy? If so, would be good to use temporary variable to hold on to that copy to avoid extra copy.
(if method returns raw value instance, does not matter)

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 6]
```diff
@@ -163,4 +165,143 @@ public static Value of(Map<Value, Value> fields) {
...
+  }
+
+  public static UUID uuid(Value value) {
+    if (value.getInnerCase() != InnerCase.UUID) {
+      throw new IllegalArgumentException("Expected uuid value");
+    }
+    ByteBuffer bytes = ByteBuffer.allocate(16);
+    value.getUuid().getValue().copyTo(bytes);
+    bytes.flip();
+    return new UUID(bytes.getLong(bytes.position()), bytes.getLong(bytes.position() + 8));
```
Comment (tatu-at-datastax) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:263]: Could just use positions `0` and `8`, since we allocated `ByteBuffer`.
That is, unless we could avoid copy altogether if `value` exposes `ByteBuffer` or allows read of 2 `long`s we need
(java-uuid-generator package would also have way to constructor `UUID` directly from `byte[]` but probably don't want to add a new dependency just for that).

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 7]
```diff
@@ -163,4 +165,143 @@ public static Value of(Map<Value, Value> fields) {
...
+  }
+
+  public static InetAddress inet(Value value) {
+    if (value.getInnerCase() != InnerCase.INET) {
+      throw new IllegalArgumentException("Expected inet value");
+    }
```
Comment (tatu-at-datastax) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:269]: Should we explicitly check the length (4 or 16 bytes) here?

File: grpc-proto/src/test/java/ValuesTest.java (hunk 2 above)
Comment (mpenick) [reply grpc-proto/src/test/java/ValuesTest.java:47]: done

File: grpc-proto/src/test/java/ValuesTest.java [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    @Test
+    public void encodeDecodeV6() throws UnknownHostException {
+      InetAddress expected = InetAddress.getByName("2001:db8::8a2e:370:7334");
+      assertThat(Values.inet(Values.of(expected))).isEqualTo(expected);
+    }
+
+    @Test
+    public void invalid() {
```
Comment (tatu-at-datastax) [inline grpc-proto/src/test/java/ValuesTest.java:290]: Should probably also test for lengths other than 4 or 16 bytes (maybe take legit Ipv4/v6 value, append one more byte).

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 3 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:181]: it returns a `long`

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 5 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:237]: it doesn't return a copy.

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 3 above)
Comment (tatu-at-datastax) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:181]: Ok then that makes sense (even if odd naming convention :) ).

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 5 above)
Comment (tatu-at-datastax) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:237]: Ok then that's fine.

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 7 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:271]: done

File: grpc-proto/src/test/java/ValuesTest.java (hunk 8 above)
Comment (mpenick) [reply grpc-proto/src/test/java/ValuesTest.java:307]: done

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 6 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java]: good catch. I wish that that `UUID(byte[])` constructor was public. Switch to using absolute positions.

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 4 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:227]: yeah, that's a good thought. I think I would like to avoid coercion mainly because it get murky fast with other types of conversions like integer types can be stored as floating point types and floating point types (and other types) can be represented as a string, etc.

Comment (tatu-at-datastax) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:227]: Yes that makes sense. And if there is need this could be relaxed at a later point.

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 6 above)
Comment (tatu-at-datastax) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java]: Right. Seems odd that constructor (or factory method) is not provided.

**Review by tatu-at-datastax — APPROVED** [review]

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 9]
```diff
@@ -163,4 +165,150 @@ public static Value of(Map<Value, Value> fields) {
         .setCollection(Collection.newBuilder().addAllElements(values).build())
         .build();
   }
+
+  public static boolean bool(Value value) {
+    if (value.getInnerCase() != InnerCase.BOOLEAN) {
+      throw new IllegalArgumentException("Expected boolean value");
```
Comment (ivansenic) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:171]: should we enrich all exception messages with the value that was provided, it might help the user?

**Review by ivansenic — APPROVED** [review]

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 9 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java]: good suggestion: https://github.com/stargate/stargate/pull/1305/commits/12fc903cccb14e4f78cae69e17dccac92eb4c73a

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 10]
```diff
@@ -163,4 +165,141 @@ public static Value of(Map<Value, Value> fields) {
...
+      throw new IllegalArgumentException(
+          String.format("Valid range for int is %d to %d", Integer.MIN_VALUE, Integer.MAX_VALUE));
+    }
+    return intValue;
+  }
+
+  public static long bigint(Value value) {
+    checkInnerCase(value, InnerCase.INT);
+
+    return value.getInt();
```
Comment (tomekl007) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:189]: shouldn't be `value.getVarint()`?

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java [hunk 11]
```diff
@@ -163,4 +165,141 @@ public static Value of(Map<Value, Value> fields) {
...
+
+    return new BigDecimal(
+        new BigInteger(value.getDecimal().getValue().toByteArray()), value.getDecimal().getScale());
+  }
+
+  public static LocalDate date(Value value) {
+    checkInnerCase(value, InnerCase.DATE);
+
+    int unsigned = value.getDate();
+    int signed = unsigned + Integer.MIN_VALUE;
```
Comment (tomekl007) [inline grpc-proto/src/main/java/io/stargate/grpc/Values.java:289]: just for curiosity, why do we need `+ Integer.MIN_VALUE` and `- Integer.MIN_VALUE;` for another way around? Can't we drop this addition/substraction?

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 10 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:189]: `.getVarint()` is for `varint` not `bigint`.

File: grpc-proto/src/main/java/io/stargate/grpc/Values.java (hunk 11 above)
Comment (mpenick) [reply grpc-proto/src/main/java/io/stargate/grpc/Values.java:289]: This is the representation of date:

```
    // CQL types: date
    // An unsigned integer representing days with Unix epoch (January, 1 1970) at 2^31.
    // Examples:
    // 0:    -5877641-06-23
    // 2^31: 1970-1-1
    // 2^32: 5881580-07-11
    uint32 date = 11;
```
